network/src/testutils/fake_socket.rs

```diff
@@ -0,0 +1,317 @@
+        self.fragmented = true;
+    }
+
+    /// reads will never return pending, but 0s
+    #[allow(dead_code)]
+    pub fn set_trailing(&mut self) {
+        self.trailing = true;
+    }
+}
+
+/// Does nothing, but looks to the caller as if write worked
+impl<'a> AsyncWrite for ReadOnlyTestSocket<'a> {
+    fn poll_write(
+        self: Pin<&mut Self>,
+        _context: &mut Context,
+        buf: &[u8],
+    ) -> Poll<io::Result<usize>> {
+        Poll::Ready(Ok(buf.len()))
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, _context: &mut Context) -> Poll<io::Result<()>> {
+        Poll::Ready(Ok(()))
+    }
+
+    /// Attempt to close the channel. Cannot Fail.
+    fn poll_close(self: Pin<&mut Self>, _context: &mut Context) -> Poll<io::Result<()>> {
+        Poll::Ready(Ok(()))
+    }
+}
+
+/// Read based on the mode set
+impl<'a> AsyncRead for ReadOnlyTestSocket<'a> {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        mut _context: &mut Context,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        // nothing left to read
+        if self.content.is_empty() {
+            if self.trailing {
+                // [0, 0, 0, 0, ...]
+                for val in buf.iter_mut() {
+                    *val = 0;
+                }
+                return Poll::Ready(Ok(buf.len()));
+            } else {
+                // EOF
+                return Poll::Ready(Ok(0));
+            }
+        }
+
+        // if something left return what's asked
+        if self.fragmented {
+            // read only one byte
+            let (read, content) = self.content.split_at(1);
+            buf[0] = read[0];
+
+            // update internal state
+            self.content = content;
+
+            // return 1 byte
+            Poll::Ready(Ok(1))
+        } else {
+            // read as much as we can
+            let to_read = std::cmp::min(buf.len(), self.content.len());
+            buf[..to_read].copy_from_slice(&self.content[..to_read]);
+
+            // update internal state
+            self.content = &self.content[to_read..self.content.len()];
+
+            // return length read
+            Poll::Ready(Ok(to_read))
+        }
+    }
+}
+
+//
+// ExposingTestSocket
+// ==================
+//
+
+pub struct ExposingTestSocket<'a> {
+    /// an in-memory socket
+    inner: MemorySocket,
+    /// useful to save what was written on the socket
+    written: Option<&'a mut Vec<u8>>,
+    /// fragment reads byte by byte
+    fragmented: bool,
+}
+
+impl<'a> ExposingTestSocket<'a> {
+    fn new(memory_socket: MemorySocket) -> Self {
+        Self {
+            inner: memory_socket,
+            written: None,
+            fragmented: false,
+        }
+    }
+
+    /// the vec passed as argument will expand to store any writes on the socket
+    pub fn save_writting(&mut self, buf: &'a mut Vec<u8>) {
+        self.written = Some(buf);
+    }
+
+    /// reads will have to be done byte by byte
+    #[allow(dead_code)]
```

Comment:
do we still need this?

Reply:
got rid of it, forgot to have testutils behind a test flag!

Reply:
I'm getting super weird errors that aren't connected at all to this PR oO

```
error[E0599]: no function or associated item named `mock` found for struct `libra_config::network_id::NetworkContext` in the current scope
   --> network/src/connectivity_manager/test.rs:178:93
    |
178 |             peer_manager::ConnectionNotification::NewPeer(peer_id, address, NetworkContext::mock()),
    |                                                                                             ^^^^ function or associated item not found in `libra_config::network_id::NetworkContext`
```

Reply:
I got those locally, and I found when I pulled all the updates and rebuilt it was fine, or at least on the build server it was fine.